include/oneapi/mkl/rng/device/detail/uniform_impl.hpp

```diff
@@ -73,25 +121,97 @@ class distribution_base<oneapi::mkl::rng::device::uniform<Type, Method>> {
+                    std::uint64_t leftover;
+
+                    sycl::vec<std::uint32_t, EngineType::vec_size> res_1 = engine.generate();
+                    sycl::vec<std::uint32_t, EngineType::vec_size> res_2 = engine.generate();
+                    sycl::vec<std::uint64_t, EngineType::vec_size> res_64;
+
+                    if constexpr (std::is_same_v<EngineType, mcg31m1<EngineType::vec_size>>) {
+                        sycl::vec<std::uint32_t, EngineType::vec_size> res_3 = engine.generate();
+
+                        for (int i = 0; i < EngineType::vec_size; i++) {
+                            res_64[i] = (static_cast<std::uint64_t>(res_3[i]) << 62) +
+                                (static_cast<std::uint64_t>(res_2[i]) << 31) + res_1[i];
+                        }
+                    }
+                    else {
+                        if constexpr (EngineType::vec_size == 3) {
+                            res_64[0] = (static_cast<std::uint64_t>(res_1[1]) << 32) +
+                                static_cast<std::uint64_t>(res_1[0]);
+                            res_64[1] = (static_cast<std::uint64_t>(res_2[0]) << 32) +
+                                static_cast<std::uint64_t>(res_1[2]);
+                            res_64[2] = (static_cast<std::uint64_t>(res_2[2]) << 32) +
+                                static_cast<std::uint64_t>(res_2[1]);
+                        } else {
+                            for (int i = 0; i < EngineType::vec_size / 2; i++) {
+                                res_64[i] = (static_cast<std::uint64_t>(res_1[2 * i + 1]) << 32) +
+                                        static_cast<std::uint64_t>(res_1[2 * i]);
+                                res_64[i + EngineType::vec_size / 2] = (static_cast<std::uint64_t>(res_2[2 * i + 1]) << 32) +
+                                                    static_cast<std::uint64_t>(res_2[2 * i]);
+                            }
+                        }
+                    }
+
+                    if (range == uint_max64)
+                        return res_64.template convert<Type>();
+
+                    for (int i = 0; i < EngineType::vec_size; i++) {
+                        leftover = res_64[i] * range;
+
+                        while (leftover < threshold) {
+                            generate_leftover<EngineType>(range, [&engine](){return engine.generate_single();},
+                                                          res_64[i], leftover);
+                        }
+
+                        res[i] = a_ + umul_hi_64(res_64[i], range);
+                    }
+
+                    return res;
+                }
             }
         }
         else {
             res = engine.generate(a_, b_);
             if constexpr (std::is_same<Method, uniform_method::accurate>::value) {
-                res = sycl::fmax(res, OutType{ a_ });
-                res = sycl::fmin(res, OutType{ b_ });
+                res = std::fmax(res, a_);
```

Comment:
why `std` func?

Reply:
I am also curious about this question. Could you share what the reason was?

Reply:
It was an intermediate solution. However, if we use the separate branch for HIP SYCL it's fine to keep `sycl`

Reply:
Ok thanks. There has been some issues with using `std` functions in SYCL kernels before so it's good that you reverted this IMO.

Reply:
That is concerning if that's the case!